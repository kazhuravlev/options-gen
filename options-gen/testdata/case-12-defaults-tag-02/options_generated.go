// Code generated by options-gen. DO NOT EDIT.
package testcase

import (
	fmt461e464ebed9 "fmt"
	"time"

	errors461e464ebed9 "github.com/kazhuravlev/options-gen/pkg/errors"
	validator461e464ebed9 "github.com/kazhuravlev/options-gen/pkg/validator"
)

type optSomeField int8

const (
	FieldSomepingPeriod  optSomeField = 0
	FieldSomename        optSomeField = 1
	FieldSomemaxAttempts optSomeField = 2
	FieldSomeeps         optSomeField = 3
)

var optSomeIsSet = [4]bool{}

type OptOptionsSetter func(o *Options)

func NewOptions(
	options ...OptOptionsSetter,
) Options {
	o := Options{}

	var empty [4]bool
	optSomeIsSet = empty

	// Setting defaults from field tag (if present)
	o.pingPeriod, _ = time.ParseDuration("3s")
	optSomeIsSet[FieldSomepingPeriod] = true
	o.name = "unknown"
	optSomeIsSet[FieldSomename] = true
	o.maxAttempts = 10
	optSomeIsSet[FieldSomemaxAttempts] = true
	o.eps = 0.0001
	optSomeIsSet[FieldSomeeps] = true

	for _, opt := range options {
		opt(&o)
	}
	return o
}

func WithSomePingPeriod(opt time.Duration) OptOptionsSetter {
	return func(o *Options) {
		o.pingPeriod = opt
		optSomeIsSet[FieldSomepingPeriod] = true
	}
}

func WithSomeName(opt string) OptOptionsSetter {
	return func(o *Options) {
		o.name = opt
		optSomeIsSet[FieldSomename] = true
	}
}

func WithSomeMaxAttempts(opt int) OptOptionsSetter {
	return func(o *Options) {
		o.maxAttempts = opt
		optSomeIsSet[FieldSomemaxAttempts] = true
	}
}

func WithSomeEps(opt float32) OptOptionsSetter {
	return func(o *Options) {
		o.eps = opt
		optSomeIsSet[FieldSomeeps] = true
	}
}

func (o *Options) Validate() error {
	errs := new(errors461e464ebed9.ValidationErrors)
	errs.Add(errors461e464ebed9.NewValidationError("pingPeriod", _validate_Options_pingPeriod(o)))
	errs.Add(errors461e464ebed9.NewValidationError("name", _validate_Options_name(o)))
	errs.Add(errors461e464ebed9.NewValidationError("maxAttempts", _validate_Options_maxAttempts(o)))
	errs.Add(errors461e464ebed9.NewValidationError("eps", _validate_Options_eps(o)))
	return errs.AsError()
}

func (o *Options) IsSet(field optSomeField) bool {
	return optSomeIsSet[field]
}

func _validate_Options_pingPeriod(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.pingPeriod, "min=100ms,max=30s"); err != nil {
		return fmt461e464ebed9.Errorf("field `pingPeriod` did not pass the test: %w", err)
	}
	return nil
}

func _validate_Options_name(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.name, "required"); err != nil {
		return fmt461e464ebed9.Errorf("field `name` did not pass the test: %w", err)
	}
	return nil
}

func _validate_Options_maxAttempts(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.maxAttempts, "min=1,max=10"); err != nil {
		return fmt461e464ebed9.Errorf("field `maxAttempts` did not pass the test: %w", err)
	}
	return nil
}

func _validate_Options_eps(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.eps, "gt=0"); err != nil {
		return fmt461e464ebed9.Errorf("field `eps` did not pass the test: %w", err)
	}
	return nil
}
