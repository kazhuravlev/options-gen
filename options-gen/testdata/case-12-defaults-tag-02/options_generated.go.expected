// Code generated by options-gen qa-version. DO NOT EDIT.

package testcase

import (
	fmt461e464ebed9 "fmt"
	"time"

	errors461e464ebed9 "github.com/kazhuravlev/options-gen/pkg/errors"
	validator461e464ebed9 "github.com/kazhuravlev/options-gen/pkg/validator"
)

type OptOptionsSetter func(o *Options)

func NewOptions(
	options ...OptOptionsSetter,
) Options {
	var o Options

	// Setting defaults from field tag (if present)

	o.pingPeriod, _ = time.ParseDuration("3s")
	o.name = "unknown"
	o.maxAttempts = 10
	o.eps = 0.0001

	for _, opt := range options {
		opt(&o)
	}
	return o
}

func WithSomePingPeriod(opt time.Duration) OptOptionsSetter {
	return func(o *Options) { o.pingPeriod = opt }
}

func WithSomeName(opt string) OptOptionsSetter {
	return func(o *Options) { o.name = opt }
}

func WithSomeMaxAttempts(opt int) OptOptionsSetter {
	return func(o *Options) { o.maxAttempts = opt }
}

func WithSomeEps(opt float32) OptOptionsSetter {
	return func(o *Options) { o.eps = opt }
}

func (o *Options) Validate() error {
	errs := new(errors461e464ebed9.ValidationErrors)
	errs.Add(errors461e464ebed9.NewValidationError("pingPeriod", _validate_Options_pingPeriod(o)))
	errs.Add(errors461e464ebed9.NewValidationError("name", _validate_Options_name(o)))
	errs.Add(errors461e464ebed9.NewValidationError("maxAttempts", _validate_Options_maxAttempts(o)))
	errs.Add(errors461e464ebed9.NewValidationError("eps", _validate_Options_eps(o)))
	return errs.AsError()
}

func _validate_Options_pingPeriod(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.pingPeriod, "min=100ms,max=30s"); err != nil {
		return fmt461e464ebed9.Errorf("field `pingPeriod` did not pass the test: %w", err)
	}
	return nil
}

func _validate_Options_name(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.name, "required"); err != nil {
		return fmt461e464ebed9.Errorf("field `name` did not pass the test: %w", err)
	}
	return nil
}

func _validate_Options_maxAttempts(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.maxAttempts, "min=1,max=10"); err != nil {
		return fmt461e464ebed9.Errorf("field `maxAttempts` did not pass the test: %w", err)
	}
	return nil
}

func _validate_Options_eps(o *Options) error {
	if err := validator461e464ebed9.GetValidatorFor(o).Var(o.eps, "gt=0"); err != nil {
		return fmt461e464ebed9.Errorf("field `eps` did not pass the test: %w", err)
	}
	return nil
}
